modules/gapi/include/opencv2/gapi/gkernel.hpp

```diff
@@ -337,29 +345,71 @@ namespace gapi {
         // Remove ALL implementations of the given API (identified by ID)
         void removeAPI(const std::string &id);
 
+        /// @private
+        // Partial includes() specialization for kernels
+        template <typename KImpl>
+        typename std::enable_if<(std::is_base_of<detail::KernelTag, KImpl>::value), bool>::type
+        includesHelper() const {
+            auto kernel_it = m_id_kernels.find(KImpl::API::id());
+            return kernel_it != m_id_kernels.end() &&
+                   kernel_it->second.first == KImpl::backend();
+        }
+
+        /// @private
+        // Partial include() specialization for kernels
+        template <typename KImpl>
+        typename std::enable_if<(std::is_base_of<detail::KernelTag, KImpl>::value), void>::type
+        includeHelper()
+        {
+            auto backend     = KImpl::backend();
+            auto kernel_id   = KImpl::API::id();
+            auto kernel_impl = GKernelImpl{KImpl::kernel()};
+            removeAPI(kernel_id);
+
+            m_id_kernels[kernel_id] = std::make_pair(backend, kernel_impl);
+        }
+
+        /// @private
+        // Partial include() specialization for transformations
+        template <typename TImpl>
+        typename std::enable_if<(std::is_base_of<detail::TransformTag, TImpl>::value), void>::type
+        includeHelper()
+        {
+            m_transformations.emplace_back(TImpl::transformation());
+        }
+
     public:
         /**
-         * @brief Returns total number of kernels in the package
-         * (across all backends included)
+         * @brief Returns total number of kernels
+         * in the package (across all backends included)
          *
          * @return a number of kernels in the package
          */
         std::size_t size() const;
 
+        /**
+         * @brief Returns vector of transformations included in the package
+         *
+         * @return vector of transformations included in the package
+         */
+        const std::vector<GTransform> &get_transformations() const;
+
         /**
          * @brief Test if a particular kernel _implementation_ KImpl is
          * included in this kernel package.
          *
          * @sa includesAPI()
          *
+         * @note cannot be applied to transformations
+         *
          * @return true if there is such kernel, false otherwise.
          */
         template<typename KImpl>
         bool includes() const
         {
-            auto kernel_it = m_id_kernels.find(KImpl::API::id());
-            return kernel_it != m_id_kernels.end() &&
-                   kernel_it->second.first == KImpl::backend();
+            static_assert(!(std::is_base_of<detail::TransformTag, KImpl>::value),
+                          "includes() cannot be applied to transformations");
```

Comment:
This check & message are not very robust.
What if later we extend our vocabulary with some new term (say, `Something` in addition to kernels and transformations) and forget about this method? Passing `Something` will raise an error that it is a transformation while it is in fact `Something`.

I'd check explicitly that `KImpl` has a kernel tag and fix text to `includes() can only be applied to kernels`.

Reply:
Will fix